eth/eth2.go

```diff
@@ -0,0 +1,325 @@
+		receipt.TransactionIndex = uint(i)
+
+		receipts[i] = new(types.Receipt)
+		*receipts[i] = *receipt
+		// Update the block hash in all logs since it is now available and not when the
+		// receipt/log of individual transactions were created.
+		for _, log := range receipt.Logs {
+			log.BlockHash = hash
+		}
+		logs = append(logs, receipt.Logs...)
+	}
+
+	block.Header().ReceiptHash = types.DeriveSha(receipts, new(trie.Trie))
+
+	return &ExecutableData{
+		Coinbase:     block.Coinbase(),
+		StateRoot:    block.Root(),
+		GasLimit:     block.GasLimit(),
+		GasUsed:      block.GasUsed(),
+		Transactions: []*types.Transaction(block.Transactions()),
+		ReceiptRoot:  block.ReceiptHash(),
+		LogsBloom:    block.Bloom().Bytes(),
+		BlockHash:    block.Hash(),
+		Difficulty:   block.Difficulty(),
+		ParentHash:   block.ParentHash(),
+	}, nil
+}
+
+// Structure described at https://hackmd.io/T9x2mMA4S7us8tJwEB3FDQ
+type NewBlockParams struct {
+	RandaoMix              common.Hash    `json:"randao_mix"`
+	Slot                   uint64         `json:"slot"`
+	Timestamp              uint64         `json:"timestamp"`
+	RecentBeaconBlockRoots []common.Hash  `json:"recent_beacon_block_roots"`
+	ExecutableData         ExecutableData `json:"executable_data"`
+}
+
+var zeroNonce [8]byte
+
+func insertBlockParamsToBlock(params NewBlockParams, number *big.Int) *types.Block {
+	header := &types.Header{
+		ParentHash:  params.ExecutableData.ParentHash,
+		UncleHash:   types.EmptyUncleHash,
+		Coinbase:    params.ExecutableData.Coinbase,
+		Root:        params.ExecutableData.StateRoot,
+		TxHash:      types.DeriveSha(types.Transactions(params.ExecutableData.Transactions), trie.NewStackTrie(nil)),
+		ReceiptHash: params.ExecutableData.ReceiptRoot,
+		Bloom:       types.BytesToBloom(params.ExecutableData.LogsBloom),
+		Difficulty:  params.ExecutableData.Difficulty,
+		Number:      number,
+		GasLimit:    params.ExecutableData.GasLimit,
+		GasUsed:     params.ExecutableData.GasUsed,
+		Time:        params.Timestamp,
+		Extra:       nil,
+		MixDigest:   common.Hash{},
+		Nonce:       zeroNonce,
+	}
+	block := types.NewBlockWithHeader(header).WithBody(params.ExecutableData.Transactions, nil /* uncles */)
+
+	return block
+}
+
+// NewBlock creates an Eth1 block, inserts it in the chain, and either returns true,
+// or false + an error. This is a bit redundant for go, but simplifies things on the
+// eth2 side.
+func (api *Eth2API) NewBlock(params NewBlockParams) (bool, error) {
+	// compute block number as parent.number + 1
+	parent := api.eth.BlockChain().GetBlockByHash(params.ExecutableData.ParentHash)
+	if parent == nil {
+		return false, fmt.Errorf("could not find parent %x", params.ExecutableData.ParentHash)
+	}
+	number := big.NewInt(0)
+	number.Add(parent.Number(), big.NewInt(1))
+
+	block := insertBlockParamsToBlock(params, number)
+	_, err := api.eth.BlockChain().InsertChainWithoutSealVerification(types.Blocks([]*types.Block{block}))
+
+	return (err == nil), err
```

Comment:
May a result object be returned here with `valid` flag in it (https://notes.ethereum.org/@n0ble/rayonism-the-merge-spec#Returns1)? The idea is to make this object extensible to allow for adding additional info in return like the reason of block invalidity (might be useful for debugging).